Document advanced binding options (#39839)
openshift/openshift-docs: modules/sbo-advanced-binding-options.adoc

// Module included in the following assemblies:
//
// * /applications/connecting_applications_to_services/binding-workloads-using-sbo.adoc

[id="sbo-advanced-binding-options_{context}"]
= Advanced binding options

> This section describes the advanced binding options, which are available only for the `binding.operators.coreos.com` API group.

Reviewer: ```suggestion
Advanced binding options are available only for the `binding.operators.coreos.com` API group.
```
I'm not sure you really need the "this section describes" phrasing, which doesn't really add any meaningful content. Your call on whether you think it's necessary.

> === Changing the binding names before projecting into workload

Reviewer: ```suggestion
=== Changing the binding names before projecting into the workload
```
Does `workload` need a definite article here? I'm not sure how it's used in this product. You decide.

Author: Yes, retaining as is.

> You can specify the rules to change the binding names in the .spec.namingStrategy attribute of the `ServiceBinding` custom resource (CR). For example, consider a Spring PetClinic sample application, which connects to the PostgreSQL database and the host and port fields of the database are exposed by the the PostgreSQL database service to use for binding. The sample application can access this binding data through the binding names.

Author: I have addressed some suggestions and have revised the content. I have ignored some suggestions to retain the technical accuracy.

.Example: Spring PetClinic sample application in the `ServiceBinding` CR
[source,yaml]
----
...
    application:
      name: spring-petclinic-rest
      group: apps
      version: v1
      resource: deployments
...
----

.Example: PostgreSQL database service in the `ServiceBinding` CR
[source,yaml]
----
...
    services:
    - group: postgres-operator.crunchydata.com
      version: v1beta1
      kind: PostgresCluster
      name: hippo
...
----

> Consider that `namingStrategy` is not defined and the binding names are projected as environment variables. Then the `host: hippo-pgbouncer` value in the backing service and the projected environment variable appears as shown in the following example:

Reviewer: ```suggestion
Consider that `namingStrategy` is not defined and the binding names are projected as environment variables. Then, the `host: hippo-pgbouncer` value in the backing service and the projected environment variable would appear as shown in the following example:
```
Use conditional for hypthetical example

> Consider that `namingStrategy` is not defined and the binding names are projected as environment variables. Then, the `host: hippo-pgbouncer` value in the backing service and the projected environment variable would appear as shown in the following example:

Reviewer: ```suggestion
If `namingStrategy` is not defined and the binding names are projected as environment variables, then the `host: hippo-pgbouncer` value in the backing service and the projected environment variable would appear as shown in the following example:
```
Very minor suggestion to make the "if, then" relationship more apparent.

.Example
----
DATABASE_HOST: hippo-pgbouncer
----
where:
[horizontal]
`DATABASE`:: Specifies the `kind` backend service.
`HOST`:: Specifies the binding name.

After applying the `POSTGRESQL_{{ .service.kind | upper }}_{{ .name | upper }}_ENV` naming strategy, the  list of custom binding names prepared by the service binding request appears as shown in the following example:

.Example
----
POSTGRESQL_DATABASE_HOST_ENV: hippo-pgbouncer
POSTGRESQL_DATABASE_PORT_ENV: 5432
----

> Following are the descriptions of the expressions defined in the `POSTGRESQL_{{ .service.kind | upper }}_{{ .name | upper }}_ENV` naming strategy:

Reviewer: ```suggestion
The following items describe the expressions defined in the `POSTGRESQL_{{ .service.kind | upper }}_{{ .name | upper }}_ENV` naming strategy:
```

> * `.name`: Refers to the binding name exposed by the backing service. In the case of previous example, the binding names are `HOST` and `PORT`.

Reviewer: ```suggestion
* `.name`: Refers to the binding name exposed by the backing service. In the case of the previous example, the binding names are `HOST` and `PORT`.
```

* `.service.kind`: Refers to the kind of service resource whose binding names are changed with the naming strategy.
* `upper`: String function used to post-process the character string while compiling the Go template string.
* `POSTGRESQL`: Prefix of the custom binding name.
* `ENV`: Suffix of the custom binding name.

Similar to the previous example, you can define the string templates in `namingStrategy` to define how each key of the binding names should be prepared by the service binding request.

=== Composing custom binding data
As an application developer, you can compose binding data under the following circumstances:

> * The backing service does not expose binding data

Reviewer: ```suggestion
* The backing service does not expose binding data.
```

* The values exposed are not available in the required format as expected by the workload.

> For example, consider a case where the backing service CR exposes the host, port, and database user as binding data, but the workload requires to consume this binding data as a connection string.

Reviewer: ```suggestion
For example, consider a case where the backing service CR exposes the host, port, and database user as binding data, but the workload requires that the binding data be consumed as a connection string.
```
